drivers/peci/peci_mchp_xec.c

```diff
@@ -0,0 +1,341 @@
+			base->CONTROL = MCHP_PECI_CTRL_FRST;
+			return -ETIMEDOUT;
+		}
+	}
+
+	/* Cleanup */
+	base->STATUS1 &= ~MCHP_PECI_STS1_EOF;
+#endif
+	return 0;
+}
+
+static int peci_xec_read(struct device *dev, struct peci_msg msg)
+{
+	int i;
+	int ret;
+	u8_t bytes_rcvd;
+	const struct peci_xec_config *config = dev->config->config_info;
+	PECI_Type *base = config->base;
+
+	/* Attempt to read data from RX FIFO */
+	bytes_rcvd = 0;
+	for (i = 0; i < (msg.read_len + PECI_FCS_LEN); i++) {
+		/* Check if data available */
+		if (!(base->STATUS2 & MCHP_PECI_STS2_RFE)) {
+			msg.read_buffer[i] = base->RD_DATA;
+			bytes_rcvd++;
+		}
+	}
+
+	/* Once transaction completes, wait for bus been idle before resetting
+	 * internal FIFOs.
+	 */
+	ret = check_bus_idle(base);
+	if (ret) {
+		return ret;
+	}
+
+	base->CONTROL = MCHP_PECI_CTRL_FRST;
+
+	return 0;
+}
+
+static int peci_xec_transfer(struct device *dev, struct peci_msg msg)
+{
+	int ret;
+	const struct peci_xec_config *config = dev->config->config_info;
+	PECI_Type *base = config->base;
+
+	/* A PECI transfer always involves a transmission attempt followed
+	 * that if successful depending on the command may or may not involve
+	 * a read operation
+	 */
+
+	ret = peci_xec_write(dev, msg);
+	if (ret) {
+		return ret;
+	}
+
+	if (msg.read_len) {
+		ret = peci_xec_read(dev, msg);
+		if (ret) {
+			return ret;
+		}
+	}
+
+	/* Check for error conditions and perform bus recovery */
+	if (base->ERROR) {
+		if (base->ERROR & MCHP_PECI_ERR_RDOV) {
+			LOG_WRN("Read buffer is not empty\n");
+		}
+
+		if (base->ERROR & MCHP_PECI_ERR_WRUN) {
+			LOG_WRN("Write buffer is not empty\n");
+		}
+
+		if (base->ERROR & MCHP_PECI_ERR_BERR) {
+			LOG_WRN("Write buffer is not empty\n");
+		}
+
+		base->ERROR = base->ERROR;
+		LOG_WRN("Transaction error\n");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_PECI_INTERRUPT_DRIVEN
+static void peci_xec_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct peci_xec_config *config = dev->config->config_info;
+	struct peci_xec_data *data = dev->driver_data;
+	PECI_Type *base = config->base;
+
+	MCHP_GIRQ_SRC(MCHP_PECI_GIRQ) = MCHP_PECI_GIRQ_VAL;
+
+	if (base->ERROR) {
+		base->ERROR = base->ERROR;
+	}
+
+	if (base->STATUS2 & MCHP_PECI_STS2_WFE) {
+		LOG_WRN("TX FIFO empty ST2:%x\n", base->STATUS2);
+		k_sem_give(&data->tx_lock);
+	}
+
+	if (base->STATUS2 & MCHP_PECI_STS2_RFE) {
+		LOG_WRN("RX FIFO full ST2:%x\n", base->STATUS2);
+	}
+}
+#endif
+
+static const struct peci_driver_api peci_xec_driver_api = {
+	.config = peci_xec_configure,
+	.enable = peci_xec_enable,
+	.disable = peci_xec_disable,
+	.transfer = peci_xec_transfer,
+};
+
+static int peci_xec_init(struct device *dev);
+
+static const struct peci_xec_config peci_xec_config = {
+	.base = (PECI_Type *) DT_PECI_MCHP_XEC_BASE_ADDRESS,
+	.irq_num = DT_PECI_MCHP_XEC_IRQ,
+};
+
+static struct peci_xec_data peci_xec_port_data;
+
+DEVICE_AND_API_INIT(peci_xec, DT_INST_0_MICROCHIP_XEC_PECI_LABEL,
```

Comment:
In my short experience with Zephyr,  I've always seen the macro DEVICE_AND_API_INIT at the very bottom